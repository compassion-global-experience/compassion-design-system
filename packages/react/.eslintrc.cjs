module.exports = {
  extends: [
    'standard-jsx',
    'standard-react',
    'plugin:jsx-a11y/recommended',
    'plugin:react/recommended',
    'plugin:react/jsx-runtime',
    'plugin:react-hooks/recommended',
    'plugin:storybook/recommended',
  ],
  env: {
    browser: true,
    es6: true,
    jest: true,
    node: false,
  },
  rules: {
    'jsx-quotes': ['error', 'prefer-double'],
  },
};
