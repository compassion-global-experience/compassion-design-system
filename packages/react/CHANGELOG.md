# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2023-04-08)

### Features

- **react:** css usage and documentation ([11e317c](https://github.com/compassion-global-experience/compassion-design-system/commit/11e317cca4ecdaa385e86d3a5571d7dca554685d))
- **react:** re-export css package ([67549a5](https://github.com/compassion-global-experience/compassion-design-system/commit/67549a58de3de6e02b58c5bdb6f4605001aae783))
- **react:** switch between themes in storybook ([79119ba](https://github.com/compassion-global-experience/compassion-design-system/commit/79119ba48ed3410eaebe9a91ccd076b05c2580c1))
