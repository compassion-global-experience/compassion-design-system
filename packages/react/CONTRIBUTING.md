# Contributing to the `react-web` Package

The `react` package is structured with a component-per-directory convention
inside the `src` directory. Each component directory generally contains 3 items:

- A **TSX file** that describes the component‚Äôs structure and behavior
- A **.stories.ts file** that illustrates most of the ways the component can
  be configured
- An **index.ts** file that exports the TSX component‚Äôs contents

## Workflow

The package uses [Storybook](https://storybook.js.org/docs/react/get-started/introduction)
to facilitate component development and documentation. Storybook uses the files
mentioned above to render components in their various states, and serves as
documentation of those components by surfacing code comments, `propTypes`, and
`defaultProps` values.

To launch Storybook, make sure you‚Äôre in the [/packages/react](/packages/react)
directory and run the command `npm run storybook`. After the project is built,
Storybook will launch in your default browser. New components or modifications
will be displayed in real time! üéâ

The configurations outlined in each component‚Äôs .stories.ts file can be seen in
Storybook‚Äôs **Docs** tab.

## Commit Messages & Versioning

Every package in the compassion-gds monorepo adheres to
[semantic versioning](https://semver.org/). In order to simplify the versioning
and release process, please be sure your commit messages adhere to the
[‚ÄúConventional Commits‚Äù pattern](https://www.conventionalcommits.org/en/v1.0.0/#summary).
You may wish to install the [commitizen/cz-cli](https://github.com/commitizen/cz-cli)
to make this near-bulletproof.

## Making a Pull Request

Once your changes are ready, open a pull request against the `development`
branch of the project, and it will be merged in as quickly as possible!
