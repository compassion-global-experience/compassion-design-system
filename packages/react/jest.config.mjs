/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/configuration
 */

export default {
  // A preset that is used as a base for Jest's configuration
  preset: 'ts-jest/presets/default-esm',

  // Automatically clear mock calls, instances, contexts and results before every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  collectCoverage: true,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // Indicates which provider should be used to instrument code for coverage
  coverageProvider: "v8",

  // The test environment that will be used for testing
  testEnvironment: "jsdom",

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  moduleNameMapper: {
    ".(css|less|scss)$": "identity-obj-proxy",
  },

  // Whether to use watchman for file crawling
  watchman: true,
};
