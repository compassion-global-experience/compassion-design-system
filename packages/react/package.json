{
  "name": "@compassion-gds/react",
  "version": "0.2.0-alpha.2",
  "description": "",
  "type": "module",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "browser": {
        "default": "./dist/esm/index.js"
      },
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js"
    },
    "./css/base.css": "./dist/css/reset.css",
    "./css/light.css": "./dist/css/cds-light.css",
    "./css/dark.css": "./dist/css/cds-dark.css",
    "./css/*": "./dist/css/*"
  },
  "scripts": {
    "dev": "rollup -c -w",
    "copy-css": "bash ./scripts/copy-css.sh",
    "prepare": "rollup -c && npm run copy-css",
    "test": "jest",
    "storybook": "storybook dev -p 6007",
    "build:storybook": "storybook build --quiet",
    "percy": "cross-env PERCY_TOKEN=$PERCY_TOKEN_REACT percy storybook ./storybook-static",
    "percy-snapshots": "npm run build:storybook && npm run percy"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/compassion-global-experience/compassion-design-system.git",
    "directory": "packages/react"
  },
  "keywords": [
    "compassion",
    "compassion-gds",
    "gds",
    "react"
  ],
  "author": "Compassion XD",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/compassion-global-experience/compassion-design-system/issues"
  },
  "homepage": "https://github.com/compassion-global-experience/compassion-design-system#readme",
  "devDependencies": {
    "@babel/core": "^7.20.12",
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@compassion-gds/core": "^0.1.0-alpha.0",
    "@rollup/plugin-commonjs": "^24.0.1",
    "@rollup/plugin-node-resolve": "^15.0.1",
    "@rollup/plugin-replace": "^5.0.2",
    "@rollup/plugin-typescript": "^11.1.0",
    "@storybook/addon-actions": "^7.0.2",
    "@storybook/addon-docs": "^7.0.2",
    "@storybook/addon-essentials": "^7.0.2",
    "@storybook/addon-interactions": "^7.0.2",
    "@storybook/addon-links": "^7.0.2",
    "@storybook/preset-react-webpack": "7.0.2",
    "@storybook/react": "7.0.2",
    "@storybook/react-webpack5": "7.0.2",
    "@storybook/testing-library": "^0.0.14-next.1",
    "@testing-library/react": "^13.4.0",
    "@types/jest": "^29.4.0",
    "@types/react": "^18.0.27",
    "babel-loader": "^8.3.0",
    "eslint": "^8.33.0",
    "eslint-config-standard-jsx": "^11.0.0",
    "eslint-config-standard-react": "^13.0.0",
    "eslint-plugin-jsx-a11y": "^6.7.1",
    "eslint-plugin-react": "^7.32.2",
    "eslint-plugin-react-hooks": "^4.6.0",
    "eslint-plugin-storybook": "^0.6.10",
    "jest": "^29.4.1",
    "jest-environment-jsdom": "^29.4.1",
    "phosphor-react": "^1.4.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "rollup": "^3.20.2",
    "rollup-plugin-dts": "^5.1.1",
    "rollup-plugin-peer-deps-external": "^2.2.4",
    "rollup-plugin-postcss": "^4.0.2",
    "storybook": "7.0.2",
    "ts-jest": "^29.0.5",
    "webpack": "^5.75.0"
  },
  "peerDependencies": {
    "react": "16.8.x || 17.x || 18.x",
    "react-dom": "16.8.x || 17.x || 18.x"
  }
}
