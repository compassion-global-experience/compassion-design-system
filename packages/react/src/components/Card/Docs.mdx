import { Markdown } from '@storybook/blocks';
import {
  Canvas,
  Controls,
  Meta,
  Stories,
  Story,
  Title,
} from '@storybook/addon-docs';

import * as CardStories from './Card.stories.tsx';

import _01_Overview from '@compassion-design-system/core/src/components/Card/docs/01_Overview.md';
import _02_Usage from '@compassion-design-system/core/src/components/Card/docs/02_Usage.md';
import _03_Image from '@compassion-design-system/core/src/components/Card/docs/03_Image.md';
import _04_Slot from '@compassion-design-system/core/src/components/Card/docs/04_Slot.md';
import _05_BestPractices from '@compassion-design-system/core/src/components/Card/docs/05_BestPractices.md';

<Meta of={CardStories} />

<Title />

<Markdown>{_01_Overview}</Markdown>
<Canvas of={CardStories.Default} sourceState="shown" withToolbar="true" />
<Controls of={CardStories.Default} />

<Markdown>{_02_Usage}</Markdown>

## Component Properties

<Markdown>{_03_Image}</Markdown>

<Markdown>{_04_Slot}</Markdown>

<Markdown>{_05_BestPractices}</Markdown>

## Examples

### Image Only

<Canvas of={CardStories.ImageOnly} />

### Content Only

<Canvas of={CardStories.ContentOnly} />
