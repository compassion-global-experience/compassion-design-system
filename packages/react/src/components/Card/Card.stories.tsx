import { Meta } from '@storybook/react';
import Card from './Card';
import Icon from '../Icon';
import imageFile from '../../stories/assets/kidsLaughing.png';

const meta: Meta<typeof Card> = {
  title: 'Components/Card',
  component: Card,
};

export default meta;

const customCardContent = (
  <div>
    <div
      style={{
        width: 72,
        height: 72,
        backgroundColor: 'lightskyblue',
        padding: 20,
        margin: 20,
        borderRadius: 36,
      }}
    >
      <Icon icon="lightbulb" description="Lightbulb" size="lg" />
    </div>
    <p>Custom content</p>
  </div>
);

export const Default = (args) => (
  <Card {...args}>
    <div>Custom content - Children rendered here</div>
  </Card>
);
Default.args = {
  imagePath: imageFile,
  imageDescription: 'Children laughing',
  title: 'Card Title',
  titleHref: '#',
  subtitle: 'Card Subtitle',
  smallText: false,
  description: 'Description',
  buttonText: 'Learn more',
  buttonOnClick: () => {
    console.log('Button clicked');
  },
  centerContent: false,
};

export const CustomContent = (args) => (
  <Card {...args}>{customCardContent}</Card>
);
CustomContent.args = {
  title: 'Card Title',
  titleHref: '#',
  subtitle: 'Card Subtitle',
  description:
    'Cards can be used to showcase different types of content, including text, images, and videos. They provide a clear and concise way to present information, while also allowing users to easily interact with the content by clicking or tapping on the card to view more details.',
  buttonText: 'Learn more',
  buttonHref: '#',
  centerContent: true,
};

export const ImageOnly = {
  args: {
    imagePath: imageFile,
  },
};

export const ContentOnly = {
  args: {
    title: 'Card Title',
    titleHref: '#',
    subtitle: 'Card Subtitle',
    description:
      'Cards can be used to showcase different types of content, including text, images, and videos. They provide a clear and concise way to present information, while also allowing users to easily interact with the content by clicking or tapping on the card to view more details.',
    buttonText: 'Learn more',
    buttonHref: '#',
    centerContent: false,
  },
};
