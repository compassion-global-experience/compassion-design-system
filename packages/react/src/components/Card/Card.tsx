import { ReactNode } from 'react';
import styles from '@compassion-design-system/core/src/components/Card/card.module.css';
import { getClasses } from '../../utils/classes';
import Button from '../Button';
import TextBox from '../TextBox';

export interface CardProps {
  imagePath?: string;
  imageDescription?: string;
  imageClassName?: string;
  title?: string;
  titleHref?: string;
  smallText?: boolean;
  subtitle?: string;
  description?: string;
  children?: ReactNode;
  buttonText?: string;
  buttonOnClick?: () => void;
  centerContent?: boolean;
}

const Card = ({
  imagePath,
  imageDescription,
  imageClassName,
  title,
  titleHref,
  smallText = false,
  subtitle,
  description,
  children,
  buttonText,
  buttonOnClick,
  centerContent = false,
}: CardProps) => {
  const isImageOnlyCard =
    !children && !title && !subtitle && !description && !buttonText;
  const titleClasses = ['cds-card__title'];
  const descriptionClasses = ['cds-card__description'];
  if (smallText) {
    titleClasses.push('small');
    descriptionClasses.push('small');
  }
  const contentClasses = ['cds-card__content'];
  if (centerContent) {
    contentClasses.push('centered');
  }
  const imageClasses = ['cds-card__image'];
  if (imageClassName) {
    imageClasses.push(imageClassName);
  }

  const CardTitle = () => {
    if (titleHref) {
      return (
        <a className={getClasses(styles, titleClasses)} href={titleHref}>
          {title}
        </a>
      );
    } else {
      return <h3 className={getClasses(styles, titleClasses)}>{title}</h3>;
    }
  };

  return (
    <div className={getClasses(styles, 'cds-card')}>
      {imagePath && (
        <img
          className={getClasses(styles, imageClasses)}
          src={imagePath}
          alt={imageDescription || 'Card image'}
        />
      )}
      {!isImageOnlyCard && (
        <div className={getClasses(styles, contentClasses)}>
          {title && <CardTitle />}
          {subtitle && (
            <h4 className={getClasses(styles, 'cds-card__subtitle')}>
              Card Subtitle
            </h4>
          )}
          {description && (
            <p className={getClasses(styles, descriptionClasses)}>
              Cards can be used to showcase different types of content,
              including text, images, and videos. They provide a clear and
              concise way to present information, while also allowing users to
              easily interact with the content by clicking or tapping on the
              card to view more details.
            </p>
          )}
          {children && <TextBox>{children}</TextBox>}
          {buttonText && (
            <div>
              <Button
                label={buttonText}
                emphasis="primary"
                size="medium"
                onClick={buttonOnClick}
              />
            </div>
          )}
        </div>
      )}
    </div>
  );
};

export default Card;
