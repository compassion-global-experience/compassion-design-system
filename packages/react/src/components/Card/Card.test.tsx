import { render } from '@testing-library/react';
import Card from './Card';

describe('Card', () => {
  it('should renders props correctly', () => {
    const { container } = render(
      <Card
        imagePath="imagePath/image.jpg"
        imageDescription="Children laughing"
        imageClassName="testImageClassName"
        title="Card Title"
        titleHref="#"
        subtitle="Card Subtitle"
        smallText={false}
        description="Description"
        buttonText="Learn more"
        buttonOnClick={() => {
          console.log('Button clicked');
        }}
        centerContent={false}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render custom content', () => {
    const { container } = render(
      <Card>
        <div>Custom content</div>
      </Card>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should center content', () => {
    const { container } = render(
      <Card description="Description" centerContent />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render small text for title and description', () => {
    const { container } = render(
      <Card title="Card Title" description="Description" smallText />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
