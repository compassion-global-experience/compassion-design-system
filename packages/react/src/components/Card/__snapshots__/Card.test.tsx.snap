// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should center content 1`] = `
<div
  class="cds-card"
>
  <div
    class="cds-card__content centered"
  >
    <p
      class="cds-card__description"
    >
      Cards can be used to showcase different types of content, including text, images, and videos. They provide a clear and concise way to present information, while also allowing users to easily interact with the content by clicking or tapping on the card to view more details.
    </p>
  </div>
</div>
`;

exports[`Card should render an image only card 1`] = `
<div
  class="cds-card"
>
  <img
    alt="Card image"
    class="cds-card__image cds-card__image--only"
    src="imagePath/image.jpg"
  />
</div>
`;

exports[`Card should render custom content 1`] = `
<div
  class="cds-card"
>
  <div
    class="cds-card__content"
  >
    <div
      class="cds-text-box cds-text-box--default cds-text-box--desktop"
    >
      <div>
        Custom content
      </div>
    </div>
  </div>
</div>
`;

exports[`Card should render small text for title and description 1`] = `
<div
  class="cds-card"
>
  <div
    class="cds-card__content"
  >
    <h3
      class="cds-card__title small"
    >
      Card Title
    </h3>
    <p
      class="cds-card__description small"
    >
      Cards can be used to showcase different types of content, including text, images, and videos. They provide a clear and concise way to present information, while also allowing users to easily interact with the content by clicking or tapping on the card to view more details.
    </p>
  </div>
</div>
`;

exports[`Card should renders props correctly 1`] = `
<div
  class="cds-card"
>
  <img
    alt="Children laughing"
    class="cds-card__image testImageClassName"
    src="imagePath/image.jpg"
  />
  <div
    class="cds-card__content"
  >
    <a
      class="cds-card__title"
      href="#"
    >
      Card Title
    </a>
    <h4
      class="cds-card__subtitle"
    >
      Card Subtitle
    </h4>
    <p
      class="cds-card__description"
    >
      Cards can be used to showcase different types of content, including text, images, and videos. They provide a clear and concise way to present information, while also allowing users to easily interact with the content by clicking or tapping on the card to view more details.
    </p>
    <div>
      <button
        class="cds-button--primary cds-button--medium"
      >
        Learn more
      </button>
    </div>
  </div>
</div>
`;
